src/components/TodoItem/TodoItem.tsx

```diff
@@ -0,0 +1,119 @@
+/* eslint-disable jsx-a11y/control-has-associated-label */
+import React, { useContext, useState } from 'react';
+import cn from 'classnames';
+import { Todo } from '../../types/todo';
+import { TodosContext } from '../../TodosContext';
+
+type Props = {
+  todo: Todo
+};
+
+export const TodoItem: React.FC<Props> = ({ todo }) => {
+  const { todos, setTodos } = useContext(TodosContext);
+  const [isEditingEnabled, setIsEditingEnabled] = useState(false);
+  const [newTitle, setNewTitle] = useState(todo.title);
+
+  const index = todos.findIndex(({ id }) => id === todo.id);
+
+  const handleRemoveTodo = () => {
+    const todosCopy = [...todos];
+
+    todosCopy.splice(index, 1);
+
+    setTodos(prev => prev.filter(({ id }) => id !== todo.id));
+  };
+
+  const handleCompleteTodo = (
+    event: React.ChangeEvent<HTMLInputElement>,
+  ) => {
+    setTodos(prev => prev.map(currentTodo => {
+      if (currentTodo.id === todo.id) {
+        return { ...currentTodo, completed: event.target.checked };
+      }
+
+      return currentTodo;
+    }));
+  };
+
+  const handleEditTodo = (
+    title: string,
+  ) => {
+    const todosCopy = [...todos];
+
+    if (title && title !== todo.title) {
+      todosCopy.splice(index, 1, { ...todo, title });
+
+      setTodos(prev => prev.map(currentTodo => {
+        if (currentTodo.id === todo.id) {
+          return { ...currentTodo, title };
+        }
+
+        return currentTodo;
+      }));
```

Comment:
https://github.com/mate-academy/react_todo-app/pull/679#discussion_r1327513903 - not fully fixed